### PR TITLE
[10.0][base_exception] Simplifies code and fix rules by domain

### PR DESCRIPTION
**Pull request by florian-dacosta**
Important note : When merging this PR, you will also need to merge PR on sale-workflow for sale_exception module, because the present sale_exception module won't work with this PR.
Sale_exception PR : https://github.com/OCA/sale-workflow/pull/855

The main goal of this PR is to fix the module and also simplify it a lot because it was becoming too complicated. (see https://github.com/OCA/server-tools/pull/1480).

It replaces https://github.com/OCA/server-tools/pull/1480

@sebastienbeau @hparfr @mourad-ehm

### CODE REVIEW

**Review by sebastienbeau — APPROVED** [review]
Code review
This new version is ok with the Specification done here  in my comment : https://github.com/OCA/server-tools/pull/1480
The code is cleanner and the domain on sale order and sale order line now work correctly without bug.

Thanks

**Comment by rousseldenis** [comment]
@Cedric-Pigeon

File: base_exception/models/base_exception.py [hunk 1]
```diff
@@ -74,14 +52,132 @@ class ExceptionRule(models.Model):
...
+
+    @api.multi
+    def _detect_exceptions_by_py_code(self, rule):
+        """
+            Find exceptions found on self.
+            If self is empty, check on all records.
+        """
+        domain = [('ignore_exception', '=', False)]
+        if self:
+            domain.append(('id', 'in', self.ids))
```
Comment (rousseldenis) [inline base_exception/models/base_exception.py]: It's better to use 

```suggestion
            domain = osv.expression.AND([domain, [('id', 'in', self.ids)]])
```

File: base_exception/models/base_exception.py [hunk 2]
```diff
@@ -74,14 +52,132 @@ class ExceptionRule(models.Model):
...
+    @api.multi
+    def _detect_exceptions_by_domain(self, rule):
+        """
+            Find exceptions found on self.
+            If self is empty, check on all records.
+        """
+        domain = rule._get_domain()
+        domain.append(('ignore_exception', '=', False))
+        if self:
+            domain.append(('id', 'in', self.ids))
```
Comment (rousseldenis) [inline base_exception/models/base_exception.py]: same as above

**Review by rousseldenis — APPROVED** [review]
Code review. LGTM

File: base_exception/models/base_exception.py (hunk 1 above)
Comment (bguillot) [inline base_exception/models/base_exception.py]: @florian-dacosta You can maybe avoid duplicate code for the domain of the search by adding a method to get the base domain

**Review by bguillot — APPROVED** [review]
Code review
My comment is not mandatory, otherwise it looks good.

Thanks for the work !

**Comment by florian-dacosta** [comment]
comments attended @bguillot @rousseldenis

**Pull request merged**